src/analysis/assert_without_msg.d

```diff
@@ -0,0 +1,186 @@
+	enum string MESSAGE = "An assert should have an explanatory message";
+
+	///
+	this(string fileName, const(Scope)* sc, bool skipTests = false)
+	{
+		super(fileName, sc, skipTests);
+	}
+
+	override void visit(const AssertExpression expr)
+	{
+		if (expr.message is null)
+			addErrorMessage(expr.line, expr.column, KEY, MESSAGE);
+		else
+		    preventCheating(expr.message);
+	}
+
+	override void visit(const FunctionCallExpression expr)
+	{
+		if (!isStdExceptionImported)
+			return;
+
+		if (expr.unaryExpression !is null &&
+			expr.unaryExpression.primaryExpression !is null &&
+			expr.unaryExpression.primaryExpression.identifierOrTemplateInstance !is null)
+		{
+			auto ident = expr.unaryExpression.primaryExpression.identifierOrTemplateInstance.identifier;
+			if (ident.text == "enforce" && expr.arguments !is null && expr.arguments.argumentList !is null)
+			{
+				if (expr.arguments.argumentList.items.length < 2)
+				    addErrorMessage(ident.line, ident.column, KEY, MESSAGE);
+                else
+                    preventCheating(expr.arguments.argumentList.items[1]);
+			}
+		}
+	}
+
+	override void visit(const SingleImport sImport)
+	{
+		static immutable stdException = ["std", "exception"];
+		if (sImport.identifierChain.identifiers.map!(a => a.text).equal(stdException))
+			isStdExceptionImported = true;
+	}
+
+	// revert the stack after new scopes
+	override void visit(const Declaration decl)
+	{
+		// be careful - ImportDeclarations don't introduce a new scope
+		if (decl.importDeclaration is null)
+		{
+			bool tmp = isStdExceptionImported;
+			scope(exit) isStdExceptionImported = tmp;
+			decl.accept(this);
+		}
+		else
+			decl.accept(this);
+	}
+
+	mixin ScopedVisit!IfStatement;
+	mixin ScopedVisit!BlockStatement;
+
+	alias visit = BaseAnalyzer.visit;
+
+private:
+	bool isStdExceptionImported;
+
+	template ScopedVisit(NodeType)
+	{
+		override void visit(const NodeType n)
+		{
+			bool tmp = isStdExceptionImported;
+			scope(exit) isStdExceptionImported = tmp;
+			n.accept(this);
+		}
+	}
+
+	void preventCheating(const ExpressionNode exprNode)
+	{
+	    import std.range : empty, walkLength;
+	    import std.uni : isWhite;
+
+		if (auto unaryExpr = cast(UnaryExpression) exprNode)
+        if (auto primaryExpr = unaryExpr.primaryExpression)
+        if (primaryExpr.primary.type == tok!"stringLiteral")
+        // quotes are part of the string literal
+        if (primaryExpr.primary.text.filter!(a => !a.isWhite).walkLength <= 2)
+            addErrorMessage(primaryExpr.primary.line, primaryExpr.primary.column, KEY, MESSAGE);
+	}
+}
+
+unittest
+{
+	import std.stdio : stderr;
+	import std.format : format;
+	import analysis.config : StaticAnalysisConfig, Check, disabledConfig;
+	import analysis.helpers : assertAnalyzerWarnings;
+
+	StaticAnalysisConfig sac = disabledConfig();
+	sac.assert_without_msg = Check.enabled;
+
+	assertAnalyzerWarnings(q{
+	unittest {
+		assert(0, "foo bar");
+		assert(0); // [warn]: %s
```

Comment:
@BBasile FWIW I test for `assert(0)` - just not in the way you would like it to be.